File: Exercises/Exercise04/Csharp.Dgemm/Csharp.Dgemm/BigIntDgemm.cs
using System;
using System.Text;
using System.Numerics;

namespace Csharp.Dgemm
{
    public class BigIntDgemm : IDgemm
    {
        private BigInteger alpha;
        private BigInteger beta;
        private BigInteger[,] matrixA;
        private BigInteger[,] matrixB;
        private BigInteger[,] matrixC;
        Reader reader;

        public BigIntDgemm(Reader reader)
        {
            this.reader = reader;
            ConvertValues();
        }

        public void ConvertValues()
        {
            try
            {
                matrixA = new BigInteger[reader.MatrixSize, reader.MatrixSize];
                matrixB = new BigInteger[reader.MatrixSize, reader.MatrixSize];
                matrixC = new BigInteger[reader.MatrixSize, reader.MatrixSize];

                alpha = BigInteger.Parse(reader.Alpha);
                beta = BigInteger.Parse(reader.Beta);
                matrixA = ConvertMatrixToBigInt(reader.MatrixA);
                matrixB = ConvertMatrixToBigInt(reader.MatrixB);
                matrixC = ConvertMatrixToBigInt(reader.MatrixC);
            }
            catch (InvalidCastException e)
            {
                Console.WriteLine(e.Message);
            }
        }

        public BigInteger[,] ConvertMatrixToBigInt(string[,] strMatrix)
        {
            BigInteger[,] matrix = new BigInteger[reader.MatrixSize, reader.MatrixSize];
            try
            {
                for (int i = 0; i < reader.MatrixSize; i++)
                {
                    for (int j = 0; j < reader.MatrixSize; j++)
                    {
                        matrix[i, j] = BigInteger.Parse(strMatrix[i, j]);
                    }
                }
            }
            catch (InvalidCastException e)
            {
                Console.WriteLine(e.Message);
            }
            return matrix;
        }

        public string[] GetAnswer()
        {
            return ConvertMatrixToString(matrixC);
        }

        public string[] ConvertMatrixToString(BigInteger[,] matrix)
        {
            string[] result = new string[reader.MatrixSize];
            for (int i = 0; i < reader.MatrixSize; i++)
            {
                StringBuilder sr = new StringBuilder();
                for (int j = 0; j < reader.MatrixSize; j++)
                {
                    sr.Append(matrix[i, j] + " ");
                }
                result[i] = sr.ToString().Trim();
            }
            return result;
        }

        public void CalculateDgemm()
        {
            matrixC = AddMatrices(MultiplyMatrixByMatrix(MultiplyScalarByMatrix(alpha, matrixA), matrixB), MultiplyScalarByMatrix(beta, matrixC));
        }


        public BigInteger[,] MultiplyMatrixByMatrix(BigInteger[,] firstMat, BigInteger[,] secondMat)
        {
            BigInteger[,] resultMatrix = new BigInteger[reader.MatrixSize, reader.MatrixSize];

            for (int m = 0; m < reader.MatrixSize; m++)
            {
                for (int n = 0; n < reader.MatrixSize; n++)
                {
                    for (int k = 0; k < reader.MatrixSize; k++)
                    {
                        resultMatrix[m, n] += firstMat[m, k] * secondMat[k, n];
                    }
                }
            }

            return resultMatrix;
        }

        public BigInteger[,] MultiplyScalarByMatrix(BigInteger scal, BigInteger[,] mat)
        {
            for (int m = 0; m < reader.MatrixSize; m++)
            {
                for (int n = 0; n < reader.MatrixSize; n++)
                {
                    mat[m, n] = mat[m, n] * scal;
                }
            }
            return mat;
        }

        public BigInteger[,] AddMatrices(BigInteger[,] firstMat, BigInteger[,] secondMat)
        {
            for (int m = 0; m < reader.MatrixSize; m++)
            {
                for (int n = 0; n < reader.MatrixSize; n++)
                {
                    firstMat[m, n] += secondMat[m, n];
                }
            }
            return firstMat;
        }
    }
}

File: Exercises/Exercise04/java_dgemm/src/benchmark/BigIntDgemm.java
package benchmark;

import java.math.BigInteger;

public class BigIntDgemm implements IDgemm{
    private BigInteger alpha;
    private BigInteger beta;
    private BigInteger[][] matrixA;
    private BigInteger[][] matrixB;
    private BigInteger[][] matrixC;
    Reader reader;

    public BigIntDgemm(Reader reader){
        this.reader = reader;
        convertValues();
    }
    @Override
    public void calculateDgemm(){
        BigInteger[][] result = addMatrices(multiplyMatrixByMatrix(multiplyScalarByMatrix(alpha, matrixA), matrixB),
                multiplyScalarByMatrix(beta, matrixC));
    }
    @Override
    public String[] getAnswer(){
        return convertMatrixToString(matrixC);
    }

    public BigInteger[][] multiplyMatrixByMatrix(BigInteger[][] firstMat, BigInteger[][] secondMat){
        BigInteger[][] resultMatrix = new BigInteger[firstMat.length][secondMat[0].length];

        for (int m = 0; m < firstMat.length; m++){
            for (int n = 0; n < secondMat[0].length; n++){
                for (int k = 0; k < secondMat.length; k++){
                    if (resultMatrix[m][n] == null){
                        resultMatrix[m][n] = firstMat[m][k].multiply(secondMat[k][n]);
                    } else {
                        resultMatrix[m][n] = resultMatrix[m][n].add(firstMat[m][k].multiply(secondMat[k][n]));
                    }

                }
            }
        }
        return resultMatrix;
    }

    public BigInteger[][] multiplyScalarByMatrix(BigInteger scal, BigInteger[][] mat){
        for (int m = 0; m < mat.length; m++){
            for (int n = 0; n < mat[0].length; n++){
                mat[m][n] = mat[m][n].multiply(scal);
            }
        }
        return mat;
    }

    public BigInteger[][] addMatrices(BigInteger[][] firstMat, BigInteger[][] secondMat){
        for (int m = 0; m < firstMat.length; m++){
            for (int n = 0; n < secondMat[0].length; n++){
                firstMat[m][n] = firstMat[m][n].add(secondMat[m][n]);
            }
        }
        return firstMat;
    }

    public void convertValues(){
        try {
            int size = reader.getMatrixSize();
            matrixA = new BigInteger[size][size];
            matrixB = new BigInteger[size][size];
            matrixC = new BigInteger[size][size];

            alpha = new BigInteger(reader.getAlpha());
            beta = new BigInteger(reader.getBeta());
            matrixA = convertMatrixToBigInt(reader.getMatrixA());
            matrixB = convertMatrixToBigInt(reader.getMatrixB());
            matrixC = convertMatrixToBigInt(reader.getMatrixC());
        } catch (NumberFormatException e) {
            System.err.println("Error: " + e.getMessage());
        }

    }

    public BigInteger[][] convertMatrixToBigInt(String[][] strMatrix){
        BigInteger[][] matrix = new BigInteger[strMatrix.length][strMatrix.length];
        try {
            for (int i = 0; i < strMatrix.length; i++){
                for (int j = 0; j < strMatrix.length; j++){
                    matrix[i][j] = new BigInteger(strMatrix[i][j]);
                }
            }
        } catch (NumberFormatException e) {
            System.err.println("Error: " + e.getMessage());
        }
        return matrix;
    }

    public String[] convertMatrixToString(BigInteger[][] matrix){
        String[] result = new String[matrix.length];
        for (int i = 0; i < matrix.length; i++){
            StringBuilder string = new StringBuilder();
            for (int j = 0; j < matrix.length; j++){
                string.append(matrix[i][j] + " ");
            }
            result[i] = string.toString().trim();
        }
        return result;
    }
}
